### ISSUE TITLE
Remaining issues with input value sanitization

### ISSUE THREAD

**Issue by KiChjang** [servo/servo#21810]
- [x] We are currently missing the value sanitization algorithm for e-mail type inputs, and the corresponding spec can be found here: https://html.spec.whatwg.org/multipage/input.html#e-mail-state-(type=email), and the implementation in the code can be found here: https://github.com/servo/servo/blob/6e844f22780a1a2b78c3a1f4bd934581144cb34e/components/script/dom/htmlinputelement.rs#L1126-L1150 (#23472)

- [ ] The following validation methods: https://github.com/servo/servo/blob/df2adebefdfa3da49f173e480fa1e56450f9bda2/components/script/dom/bindings/str.rs#L273-L301 need to be rewritten, as I have communicated with WHATWG before and found out that successfully parsing the string does not mean a successful validation. We'll need to go through each of the input types in the spec and look at their validation algorithm to see how it should look like instead:
https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-date-string
https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-month-string
https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-week-string
https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-floating-point-number

- [ ] Remove the catch-all case here https://github.com/servo/servo/blob/6e844f22780a1a2b78c3a1f4bd934581144cb34e/components/script/dom/htmlinputelement.rs#L1188
so that we don't forget about implementing value sanitizations on any of the input types again.

**Comment by chansuke** [comment]
@highfive: assign me

**Comment by tigercosmos** [comment]
@KiChjang I am curious about the statement, "successfully parsing the string does not mean a successful validation".
Take `date` type for example, three points in `valid date string` are all in `parse a date component`, and `parse a date component` even has more steps. In this case, is the statement true or false?

**Comment by KiChjang** [comment]
@tigercosmos I found the relevant IRC chat record that prompted this: https://freenode.logbot.info/whatwg/20180117#c1402104
TL;DR floating point strings are allowed to have whitespaces in front of them during parsing, in which case they are ignored, but the validation algorithm would report invalid on floating point strings that have extra whitespace in front.

**Comment by tigercosmos** [comment]
BTW, WPT tests passed before with the old implementaion, so WPT tests should also be updated to make the behavior correct.

**Comment by chansuke** [comment]
@KiChjang I'm now rewriting the `is_valid_date_string`  but blocking with `A U+002D HYPHEN-MINUS character (-)` part.
Could you perhaps give me some suggestion about the way to detect the `-` character.Thanks.

**Comment by KiChjang** [comment]
@chansuke How exactly are you scanning for characters? Is it through the `chars()` iterator or the `bytes()` iterator? Either way, you can check for it using the `'-'` char literal (or `b'-'` if you're using `bytes()`).

**Comment by KiChjang** [comment]
@chansuke Was this completed?

**Comment by CYBAI** [comment]
@chansuke Is this still in progress?

**Comment by chansuke** [comment]
@KiChjang @CYBAI Sorry for a late reply.The status is WIP.I couldn't take a time in this 2months.I will handle this in this weekend.

**Comment by CYBAI** [comment]
Remove `C-assigned` label due to inactivity. Also, @chansuke feel free to get back to this when you have time if this is not fixed yet.

**Comment by ffwff** [comment]
Can I work on email value sanitization?

**Comment by CYBAI** [comment]
@ffwff Go ahead! :)

**Comment by glowe** [comment]
@highfive: assign me. I'd like to handle the two remaining items on this issue.

**Comment by glowe** [comment]
I looked at the specifications and their corresponding parsing specifications:

- https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-date-string
- https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-month-string
- https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-week-string

It seems like continuing to rely on the parsing implementations will still be correct. Can someone verify my assessment?

If we do decide that the work still needs to be done, would it be acceptable to use the `regex` crate? It's already used in other parts of `servo` and would allow us to succinctly express the algorithms.

I can't tell whether `is_valid_floating_point_number` needs to be updated yet. If it turns out it does, could we also use the `regex` crate for this?

**Comment by jdm** [comment]
Sorry, I missed these questions. Looking at those three links, I agree that those parsing specifications appear to match the corresponding validation specification. I don't think it's worth rewriting the validation routines unless there is motion in the specification to change them, or we learn of a case that is not actually covered for some reason.

As for floating point numbers, it appears that parsing includes skipping whitespace, and validation does not. Using regex sounds like a reasonable choice to me.

**Comment by glowe** [comment]
> As for floating point numbers, it appears that parsing includes skipping whitespace, and validation does not. Using regex sounds like a reasonable choice to me.

Thanks. It seems like we actually have a bug in `is_valid_floating_point_number`. It currently rejects floating point strings that start with a space, but allows leading tabs, newlines, etc. So this function needs updating.
